Return value support for BYOB

cgillum opened the pull request:
This is to resolve https://github.com/Azure/azure-webjobs-sdk/issues/1104.

Phase 1 simply allows functions in WebJobs to have a return value. Previously we supported `void` and `Task`. With this change, we also support `TReturnValue` and `Task<TReturnValue>`.

Example functions I'm using for testing:

```csharp
public static int QueueTriggerWithReturnValue(
    [QueueTrigger("test")] string message)
{
    Console.WriteLine("Processed message: " + message);
    return 5;
}

public static Task<int> QueueTriggerWithReturnValueAsync(
    [QueueTrigger("test")] string message)
{
    Console.WriteLine("Processed message: " + message);
    return Task.FromResult(5);
}
```



Phase 2 (TBD) is to connect the return value back to the trigger binding.

dnfclas commented:
@cgillum,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

MikeStall commented on src/Microsoft.Azure.WebJobs.Host/Executors/FunctionInvokerFactory.cs:
```diff
@@ -27,25 +28,42 @@ public static IFunctionInvoker Create(MethodInfo method, IJobActivator activator
+            else if (typeof(Task).IsAssignableFrom(returnType))
+            {
+                returnType = returnType.GetGenericArguments()[0];
+            }
```
Nit: can this logic be in TypeUtility next to your related helper?

cgillum replied:
Looks like I can replace this logic with TypeUtility.TryGetReturnType.

MikeStall commented on src/Microsoft.Azure.WebJobs.Host/Indexers/FunctionIndexer.cs:
```diff
@@ -198,6 +205,34 @@ private bool HasJobAttribute(CustomAttributeData attributeData)
+                if (bindingDataContract != null && bindingDataContract.TryGetValue(ReturnParamName, out Type triggerReturnType))
+                {
+                    // CONSIDER: Special casing for JObject and related types for cross-language support?
+                    if (!triggerReturnType.IsAssignableFrom(methodReturnType))
```
>if (!triggerReturnType.IsAssignableFrom(methodReturnType)) [](start = 20, length = 58)

We probably need to go through the converter managers here, which  would handle this.   For now, the trigger can just provide a contract of type  Object, which makes this moot.

fabiocav commented on src/Microsoft.Azure.WebJobs.Host/Executors/TriggeredFunctionData.cs:
```diff
@@ -24,6 +24,6 @@ public class TriggeredFunctionData
         /// Optional handler function for processing the invocation.
         /// </summary>
-        public Func<Func<Task>, Task> InvokeHandler { get; set; }
+        public Func<Func<Task<object>>, Task<object>> InvokeHandler { get; set; }
```
Since we intend to have these changes in 2.x, we can't make public API breaking changes.

cgillum commented:
I've rebased the PR on top of the .NET Core changes in the dev branch. As a result, I had to do a force push on the PR branch which unfortunately invalidated previous review comments.

@fabiocav I have reverted the breaking change you had asked about previously. The result is we simply don't support return values when going through the InvokeHandler path. Durable Functions actually doesn't need this in it's InvokeHandler case, so it's not a problem to leave out this support (for now).

MikeStall commented:
:shipit:
